### PR TITLE
add parallel for all tasks

### PR DESCRIPTION
**Pull request by aloeAlika**

### CODE REVIEW

File: task2_11.04/src/main/scala/AppStatistic.scala [hunk 1]
```diff
@@ -16,6 +11,6 @@ object AppStatistic {
     meanOverallAndCountOfVotedUsersByApp.toList.foreach(r => rows += List(r._1, r._2._1.toString, r._2._2.toString))
```
[the remark this answers is not in the document]
Comment (aloeAlika) [reply task2_11.04/src/main/scala/AppStatistic.scala]: fixed it

File: task2_11.04/src/main/scala/AppStatistic.scala [hunk 2]
```diff
@@ -2,11 +2,6 @@ import scala.collection.mutable.ListBuffer
 import Main.{addPrefix, writeCsvFile}
 
 object AppStatistic {
-  def main(seqOfReviewsFromJson: Seq[Review]): Unit = {
-    val meanOverallAndCountOfVotedUsersByApp = infoAboutApps(seqOfReviewsFromJson)
-    //writeAppStatisticToCsv(meanOverallAndCountOfVotedUsersByApp)
-  }
- // def infoAboutApps(seqOfReviewsFromJson: Seq[Review]) = seqOfReviewsFromJson.groupBy(_.asin).mapValues(r => (r.map(_.overall).sum / r.size,r.size))
 
   def infoAboutApps(seqOfReviewsFromJson: Seq[Review]) = seqOfReviewsFromJson.groupBy(_.asin).mapValues(r => (r.map(_.overall)))
```
[the remark this answers is not in the document]
Comment (aloeAlika) [reply task2_11.04/src/main/scala/AppStatistic.scala]: done

File: task2_11.04/src/main/scala/Main.scala [hunk 3]
```diff
@@ -18,30 +18,62 @@ object Main {
...
-    //writeMostlyLikedCommentToCsv(mostlyLikedReview)
-    //val shortestTimeBetweenRatingByUsers = timeBetweenRating(seqOfReviewsFromJson)
-    //writingTimeBetweenRatingToCsv(shortestTimeBetweenRatingByUsers)
-    //val reviewWithHelpllesComment = appWhereHelpfullessComment(seqOfReviewsFromJson)
-    //writeAppWhereHelpfullessCommentIsToCsv(reviewWithHelpllesComment)
-    //AppStatistic.main(seqOfReviewsFromJson)
-    //WordsStatistic.popularWordsToCsv(seqOfReviewsFromJson)
+  }
+
+  def parallelForTask4c(filename_1: String, filename_2: String, filename_3: String):  Future[List[HashMap[String, (Long, Int, Int)]]] =
```
[the remark this answers is not in the document]
Comment (aloeAlika) [reply task2_11.04/src/main/scala/Main.scala]: done

File: task2_11.04/src/main/scala/Main.scala [hunk 4]
```diff
@@ -176,10 +208,9 @@ object Main {
     mostlyLikedReview.foreach(r => rows += List(r.helpful.head.toString, r.reviewText))
```
[the remark this answers is not in the document]
Comment (aloeAlika) [reply task2_11.04/src/main/scala/Main.scala]: done

File: task2_11.04/src/main/scala/WordsStatistic.scala [hunk 5]
```diff
@@ -25,11 +20,11 @@ object WordsStatistic {
     listsOfPopularWords._1.foreach(r => rows1 += List(r._1, r._2.toString))
```
[the remark this answers is not in the document]
Comment (aloeAlika) [reply task2_11.04/src/main/scala/WordsStatistic.scala]: fixed it

File: task2_11.04/src/main/scala/WordsStatistic.scala [hunk 6]
```diff
@@ -25,11 +20,11 @@ object WordsStatistic {
...
     var endRows: List[List[String]] = rows1.toList
     endRows = addPrefix(endRows)
-    writeCsvFile("words-stats1.csv", header, endRows)
+    writeCsvFile("result/words-stats1.csv", header, endRows)
     val rows2: ListBuffer[List[String]] = ListBuffer()
     listsOfPopularWords._2.foreach(r => rows2 += List(r._1, r._2.toString))
```
[the remark this answers is not in the document]
Comment (aloeAlika) [reply task2_11.04/src/main/scala/WordsStatistic.scala]: fixed it

File: task2_11.04/src/main/scala/AppStatistic.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (sleshJdev) [reply task2_11.04/src/main/scala/AppStatistic.scala]: thanks

File: task2_11.04/src/main/scala/Main.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (sleshJdev) [reply task2_11.04/src/main/scala/Main.scala]: Could you please attach a link to appropriate commit

File: task2_11.04/src/main/scala/Main.scala [hunk 7]
```diff
@@ -18,30 +18,62 @@ object Main {
   implicit val ec: ExecutionContext = ExecutionContext.global
 
   def main(args: Array[String]): Unit = {
```
[the remark this answers is not in the document]
Comment (sleshJdev) [reply task2_11.04/src/main/scala/Main.scala]: Here is an example:
```

object Hello extends App {
    println("Hello, world")
}
```

so don't need main method anymore

File: task2_11.04/src/main/scala/WordsStatistic.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (sleshJdev) [reply task2_11.04/src/main/scala/WordsStatistic.scala]: Thanks

File: task2_11.04/src/main/scala/WordsStatistic.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (sleshJdev) [reply task2_11.04/src/main/scala/WordsStatistic.scala]: good job

[the remark this answers is not in the document]
Comment (sleshJdev) [reply task2_11.04/src/main/scala/WordsStatistic.scala]: commit link?

File: task2_11.04/src/main/scala/AppStatistic.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (aloeAlika) [reply task2_11.04/src/main/scala/AppStatistic.scala]: @sleshJdev https://github.com/aloeAlika/big_data_intern/commit/8858cefb06f45d274edda50595172e01119489f2

File: task2_11.04/src/main/scala/AppStatistic.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (aloeAlika) [reply task2_11.04/src/main/scala/AppStatistic.scala]: @sleshJdev https://github.com/aloeAlika/big_data_intern/commit/b0c0425a782021dc43f5301f80431b6b5ffc0686

File: task2_11.04/src/main/scala/Main.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (aloeAlika) [reply task2_11.04/src/main/scala/Main.scala]: @sleshJdev https://github.com/aloeAlika/big_data_intern/commit/b64de089d14f8613844265a16e5b178a4c06575b

File: task2_11.04/src/main/scala/WordsStatistic.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (aloeAlika) [reply task2_11.04/src/main/scala/WordsStatistic.scala]: https://github.com/aloeAlika/big_data_intern/commit/8858cefb06f45d274edda50595172e01119489f2

File: task2_11.04/src/main/scala/WordsStatistic.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (aloeAlika) [reply task2_11.04/src/main/scala/WordsStatistic.scala]: https://github.com/aloeAlika/big_data_intern/commit/8858cefb06f45d274edda50595172e01119489f2

File: task2_11.04/src/main/scala/Main.scala [hunk 8]
```diff
@@ -18,30 +18,62 @@ object Main {
...
-    val m = Map[String, Seq[Float]]()
-    val c = scala.collection.immutable.HashMap[String, Seq[Float]](m.toList:_*)
-    val b = scala.collection.immutable.HashMap[String, Seq[Float]]()
-    b.merged(c) {
-      case ((k1, v1), (k2, v2)) => (k1, v1 ++ v2)
-    }
     val filename_1 = "/home/aliaksandra/IdeaProjects/big_data_intern/task2_11.04/src/main/resources/sources/App_reviews_1.json"
```
[the remark this answers is not in the document]
Comment (aloeAlika) [reply task2_11.04/src/main/scala/Main.scala]: done 
https://github.com/aloeAlika/big_data_intern/commit/4c34475e235701d377e4ba282e589341a780b6aa

File: task2_11.04/src/main/scala/Main.scala [hunk 9]
```diff
@@ -18,30 +18,62 @@ object Main {
...
     val filename_2 = "/home/aliaksandra/IdeaProjects/big_data_intern/task2_11.04/src/main/resources/sources/App_reviews_2.json"
```
[the remark this answers is not in the document]
Comment (aloeAlika) [reply task2_11.04/src/main/scala/Main.scala]: done 
https://github.com/aloeAlika/big_data_intern/commit/df0764f39fd85adc033f4950c7cf8acab0328c56

File: task2_11.04/src/main/scala/Main.scala [hunk 10]
```diff
@@ -18,30 +18,62 @@ object Main {
...
+      }
+      case Failure(e) => println("Failure")
+    }
+
+  def futureProcessingTask6(path: String): (HashMap[String, Int],HashMap[String, Int])= {
+    println("Future path: " + path)
+    WordsStatistic.getMapsOfPopularWords(readingFromJsonToSeq(path))
   }
 
   def parallelForTask5(filename_1: String, filename_2: String, filename_3: String):  Future[List[Map[String, Seq[Float]]]] =
```
[the remark this answers is not in the document]
Comment (aloeAlika) [reply task2_11.04/src/main/scala/Main.scala]: no more warnings :) 
https://github.com/aloeAlika/big_data_intern/commit/df0764f39fd85adc033f4950c7cf8acab0328c56

File: task2_11.04/src/main/scala/Main.scala [hunk 11]
```diff
@@ -18,30 +18,62 @@ object Main {
...
+    println("Future path: " + path)
+    scala.collection.immutable.HashMap(timeBetweenRating(readingFromJsonToSeq(path)).toList:_*)
+  }
+  def parallelForTask6(filename_1: String, filename_2: String, filename_3: String):  Future[List[(HashMap[String, Int],HashMap[String, Int])]] =
+    Future.sequence(List(filename_1, filename_2, filename_3)
+      .map(path => Future(futureProcessingTask6(path)))
+    ) andThen {
+      case Success(values) => {
+        val pos = values.apply(0)._1.merged(values.apply(1)._1) {
+          case ((k1, v1), (k2, v2)) => (k1, v1 + v2)
```
[the remark this answers is not in the document]
Comment (aloeAlika) [reply task2_11.04/src/main/scala/Main.scala]: done 
https://github.com/aloeAlika/big_data_intern/commit/2f60cf43da3b60d21661c3cb11b78360c92ef681

File: task2_11.04/src/main/scala/Main.scala [hunk 12]
```diff
@@ -18,30 +18,62 @@ object Main {
...
-    //WordsStatistic.popularWordsToCsv(seqOfReviewsFromJson)
+  }
+
+  def parallelForTask4c(filename_1: String, filename_2: String, filename_3: String):  Future[List[HashMap[String, (Long, Int, Int)]]] =
+    Future.sequence(List(filename_1, filename_2, filename_3)
+      .map(path => Future(futureProcessingTask4c(path)))
+    ) andThen {
+      case Success(values) => {
+        val w = values.fold(HashMap())((l,r) => l.merged(r) {
+          case ((k1, v1), (k2, v2)) => (k1, {
```
[the remark this answers is not in the document]
Comment (aloeAlika) [reply task2_11.04/src/main/scala/Main.scala]: > Please, refactor this code to make it shorted or more readable by extracting into separate method with semantic name or in some different way. It's up to you.

fixed it 
https://github.com/aloeAlika/big_data_intern/commit/2f60cf43da3b60d21661c3cb11b78360c92ef681

File: task2_11.04/src/main/scala/AppStatistic.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (sleshJdev) [reply task2_11.04/src/main/scala/AppStatistic.scala]: Thanks, _seqOfReviews_ prefix _seqOf_ is also redundant. Just _reviews_ is fine

File: task2_11.04/src/main/scala/Main.scala [hunk 13]
```diff
@@ -18,30 +18,62 @@ object Main {
...
+          case ((k1, v1), (k2, v2)) => (k1, v1 + v2)
+        }.merged(values.apply(2)._1) {
+          case ((k1, v1), (k2, v2)) => (k1, v1 + v2)
+        }.toList
+        val neg = values.apply(0)._2.merged(values.apply(1)._2) {
+          case ((k1, v1), (k2, v2)) => (k1, v1 + v2)
+        }.merged(values.apply(2)._2) {
+          case ((k1, v1), (k2, v2)) => (k1, v1 + v2)
+        }.toList
+        WordsStatistic.writePopularWordsToCsv(pos, neg)
```
[the remark this answers is not in the document]
Comment (aloeAlika) [reply task2_11.04/src/main/scala/Main.scala]: done 
https://github.com/aloeAlika/big_data_intern/commit/9437b12524b610ebd2d21b54d778181212a837ad

File: task2_11.04/src/main/scala/Main.scala (hunk 12 above)
[the remark this answers is not in the document]
Comment (aloeAlika) [reply task2_11.04/src/main/scala/Main.scala]: > Overall app comment: let's think about metric computations as about business logic. So, could you please extract all related methods into separate core package(module, whatever) and use it here as dependency. The point is that there shouldn't mix up of you business logic and writing into file or another side effects.



> Also, please create such module for IO operations. Let's try to keep Main class short and concise. Now it looks overloaded that scares you.

done 
https://github.com/aloeAlika/big_data_intern/commit/9437b12524b610ebd2d21b54d778181212a837ad

**Pull request closed without merging**